gensim/models/ldaseqmodel.py

```diff
@@ -0,0 +1,1222 @@
+
+
+    T = sslm.num_time_slices
+
+    obs = sslm.obs[word]
+    fwd_variance = sslm.fwd_variance[word]
+
+    mean = sslm.mean[word]
+    fwd_mean = sslm.fwd_mean[word]
+
+    # forward 
+    fwd_mean[0] = 0
+    for t in range(1, T + 1):
+        w = sslm.obs_variance / (fwd_variance[t - 1] + chain_variance + sslm.obs_variance)
+        fwd_mean[t] = w * fwd_mean[t - 1] + (1 - w) * obs[t - 1]
+
+    # backward pass
+    mean[T] = fwd_mean[T]
+    for t in range(T - 1, -1, -1):
+        if chain_variance == 0.0:
+            w = 0.0
+        else:
+            w = chain_variance / (fwd_variance[t] + chain_variance)
+        mean[t] = w * fwd_mean[t] + (1 - w) * mean[t + 1]
+
+    return mean, fwd_mean
+
+
+def update_phi(doc, time, lda_post):
+
+    """
+    Update variational multinomial parameters, based on a document and a time-slice.
+    This is done based on the original Blei-LDA paper, where:
+    log_phi := beta * exp(Ψ(gamma)), over every topic for every word.
+    
+    TODO: incorporate lee-sueng trick used in **Lee, Seung: Algorithms for non-negative matrix factorization, NIPS 2001**.
+    """
+
+    K = lda_post.lda.num_topics
+    N = lda_post.doc.nterms
+
+    dig = numpy.zeros(K)
+
+    for k in range(0, K):
+        dig[k] = digamma(lda_post.gamma[k])
+
+    for n in range(0, N):
+        w = lda_post.doc.word[n]
+        for k in range(0, K):
+            lda_post.log_phi[n][k] = dig[k] + lda_post.lda.topics[w][k]
+
+        log_phi_row = lda_post.log_phi[n]
+        phi_row = lda_post.phi[n]
+
+        # log normalize
+        v = log_phi_row[0]
+        for i in range(1, len(log_phi_row)):
+            v = numpy.logaddexp(v, log_phi_row[i])
+
+        for i in range(0, len(log_phi_row)):
+            log_phi_row[i] = log_phi_row[i] - v
+
+        for k in range(0, K):
+            phi_row[k] = numpy.exp(log_phi_row[k])
+
+        lda_post.log_phi[n] = log_phi_row
+        lda_post.phi[n] = phi_row
+
+    return lda_post.phi, lda_post.log_phi
+
+def update_gamma(lda_post):
+
+    """
+    update variational dirichlet parameters as described in the original Blei LDA paper:
+    gamma = alpha + sum(phi), over every topic for every word.
+
+    """
+
+    K = lda_post.lda.num_topics
+    N = lda_post.doc.nterms
+
+    lda_post.gamma = numpy.copy(lda_post.lda.alpha)
+
+    for n in range(0, N):
+        phi_row = lda_post.phi[n]
+        count = lda_post.doc.count[n]
+
+        for k in range(0, K):
+            lda_post.gamma[k] += phi_row[k] * count
+
+    return lda_post.gamma
+
+def compute_expected_log_prob(sslm):
+
+    """
+    Compute the expected log probability given values of m.
+    The appendix describes the Expectation of log-probabilities in equation 5 of the DTM paper;
+    The below implementation is the result of solving the equation and is as implemented in the original Blei DTM code.
+    """
+
+    W = sslm.vocab_len
+    T = sslm.num_time_slices
+    for t in range(0, T):
+        for w in range(0, W):
```

Comment:
It can't be for this right? At best I can do for `w`. And by vectorise here do you mean exactly? I still need values of `w` and `t` to access my `mean` values right?